File: typescript/LWTSD/ResourceTypes/ResourceDateTime.ts
var xml = require('../../node-xmpp/packages/xml');

import { LWTSD } from '../index';
import { SimplifiedType } from '../SimplifiedType';
import { ResourceDescription } from '../ResourceDescription'
import { ResourceRestriction } from '../ResourceRestriction';
import { Resource } from './Resource'

export class ResourceDateTime extends Resource
{
    public readonly HistoricalValues :  Array< [Date, Date] > = new Array< [Date, Date] >();

    InnerValue : Date;

    public GetValue() : Date
    {
        return this.InnerValue;
    }

    public SetValue(Value : Date) : void
    {
        if(this.InnerMinExclusive && Value < this.InnerMinExclusive)
            throw new Error("Value not within limits");
        if(this.InnerMaxExclusive && Value > this.InnerMaxExclusive)
            throw new Error("Value not within limits");

        this.InnerValue = Value;
        this.ModifiedAt = new Date();
    }

    public constructor(Element? : any)
    {
        super(Element);

        let points = Element.getChildren("point");
        for (let i = 0; i < points.length; i++)
        {
            let it = points[i];
            let TimeStamp = new Date(it.attrs.timestamp);
            let PointValue : Date = new Date(it.text());
            this.HistoricalValues.push( [TimeStamp, PointValue] );
            if (TimeStamp > this.ModifiedAt)
            {
                this.ModifiedAt = TimeStamp;
                this.InnerValue = PointValue;
            }
        }
    }

    public GetPoint() : any
    {
        let RVal = xml("point", { /*"xmlns": LWTSD.Namespace */}, this.GetValue());
        RVal.attrs.timestamp = this.ModifiedAt;
        return RVal;
    }

    public GetWriteValue() : string
    {
        return this.GetValue().toISOString();
    }

    public GetDescription() : ResourceDescription
    {
        return super.GetDescriptionFromSimpleType(SimplifiedType.DateTime);
    }

    public LoadFromWrite(WriteElement : any) : void
    {
        this.SetValue( WriteElement.text() );
    }
}

File: NET/LWTSD/ResourceTypes/ResourceDateTime.cs
using System;
using System.Xml;
using System.Xml.Linq;
using System.Collections.Generic;

namespace LWTSD.ResourceTypes
{
	public class ResourceDateTime : Resource
	{
		public readonly List<Tuple<DateTime, DateTime>> HistoricalValues = null;

		DateTime InnerValue;
		public DateTime Value
		{
			get
			{
				return InnerValue;
			}

			set
			{
				if (MinExclusive != null)
					if (value < MinExclusive)
						throw new InvalidOperationException("Value not within limits");
				if (MaxExclusive != null)
					if (value > MaxExclusive)
						throw new InvalidOperationException("Value not within limits");

				InnerValue = value;

				ModifiedAt = DateTime.UtcNow;
			}
		}

		public DateTime? MinExclusive
		{
			get
			{
				if (InnerMinExclusive == null)
					return null;

				return (DateTime) InnerMinExclusive;
			}

			set
			{
				InnerMinExclusive = value;
			}
		}
		public DateTime? MaxExclusive
		{
			get
			{
				if (InnerMaxExclusive == null)
					return null;

				return (DateTime) InnerMaxExclusive;
			}

			set
			{
				InnerMaxExclusive = value;
			}
		}

		public override XElement GetPoint()
		{
			XElement Point = base.GetPoint();
			Point.Value = XmlConvert.ToString(Value, XmlDateTimeSerializationMode.Utc);
			return Point;
		}

		public override ResourceDescription GetDescription()
		{
			return base.GetDescription(SimplifiedType.DateTime);
		}

		public override void LoadFromWrite(XElement WriteElement)
		{
			this.Value = XmlConvert.ToDateTime(WriteElement.Value, XmlDateTimeSerializationMode.Utc);
		}

		public override string GetWriteValue()
		{
			return Value.ToString();
		}

		public ResourceDateTime(XElement Element)
			: base(Element)
		{
			ModifiedAt = DateTime.MinValue;
			foreach (var it in Element.Elements(LWTSD.Namespace + "point"))
			{
				if (HistoricalValues == null)
					HistoricalValues = new List<Tuple<DateTime, DateTime>>();
				DateTime TimeStamp = XmlConvert.ToDateTime(it.Attribute("timestamp").Value,
														   XmlDateTimeSerializationMode.Utc);
				DateTime PointValue = XmlConvert.ToDateTime(it.Value, XmlDateTimeSerializationMode.Utc);
				HistoricalValues.Add(new Tuple<DateTime, DateTime>(TimeStamp, PointValue));

				if (TimeStamp > ModifiedAt)
				{
					ModifiedAt = TimeStamp;
					InnerValue = PointValue;
				}

			}
		}

		public ResourceDateTime()
		{
		}

	}
}
